### PR TITLE
Fix several double-free errors in InternalMemMgr

### PR DESCRIPTION
**Pull request by samikhawaja**
The Vulkan CTS has a number of OOM tests that caused double-free errors in InternalMmeMgr. On allocation failure the cleanup code doesn't check whether the allocations succeeded before doing deallocations.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

File: icd/api/internal_mem_mgr.cpp [hunk 1]
```diff
@@ -473,8 +473,12 @@ VkResult InternalMemMgr::CreateMemoryPoolAndSubAllocate(
         Pal::Result palResult = pOwnerList->PushFront(newPool);
         result = PalToVkResult(palResult);
         VK_ASSERT(result == VK_SUCCESS);
```
Comment (JonCampbell407) [inline icd/api/internal_mem_mgr.cpp:475]: Get rid of this assert since we now handle it not being VK_SUCCESS.

File: icd/api/internal_mem_mgr.cpp [hunk 2]
```diff
@@ -495,33 +499,37 @@ VkResult InternalMemMgr::CreateMemoryPoolAndSubAllocate(
...
     }
-    else
+
+    if (newPool.pBuddyAllocator == nullptr)
     {
-        auto it = pOwnerList->Begin();
-        bool needEraseFromOwnerList = pOwnerList->NumElements() > 0 ?
-            (it.Get()->groupMemory.PalMemory(DefaultDeviceIndex) ==
-             pInternalMemory->groupMemory.PalMemory(DefaultDeviceIndex)) : false;
+        return result;
```
Comment (JonCampbell407) [inline icd/api/internal_mem_mgr.cpp:507]: Avoid early return.

**Review by JonCampbell407 — CHANGES_REQUESTED** [review]

File: icd/api/internal_mem_mgr.cpp (hunk 1 above)
Comment (samikhawaja) [reply icd/api/internal_mem_mgr.cpp]: Done

File: icd/api/internal_mem_mgr.cpp (hunk 2 above)
Comment (samikhawaja) [reply icd/api/internal_mem_mgr.cpp]: Done

File: icd/api/internal_mem_mgr.cpp [hunk 3]
```diff
@@ -496,30 +499,30 @@ VkResult InternalMemMgr::CreateMemoryPoolAndSubAllocate(
...
+            FreeBaseGpuMem(pInternalMemory);
+
+            // Remove `pInternalMemory` from `pOwnerList`.
+            auto it = pOwnerList->Begin();
+            VK_ASSERT(pInternalMemory == it.Get());
             pOwnerList->Erase(&it);
+            pInternalMemory = nullptr;
         }
+
+        VK_ASSERT(newPool.pBuddyAllocator != nullptr);
```
Comment (JonCampbell407) [inline icd/api/internal_mem_mgr.cpp:521]: We don't need this assert since we are inside a test for it.

**Review by JonCampbell407 — CHANGES_REQUESTED** [review]

File: icd/api/internal_mem_mgr.cpp (hunk 3 above)
Comment (samikhawaja) [reply icd/api/internal_mem_mgr.cpp]: Done.

**Review by JonCampbell407 — APPROVED** [review]

**Comment by JaxLinAMD** [comment]
ok to test

**Pull request merged**